trailmap/pathway_analysis.py

```diff
@@ -0,0 +1,319 @@
+            "DiGraph. Proceeding with caution, consider the results at " +
+            "your own risk\n")
+            H = nx.DiGraph(G)
+        elif multiedges == False:
+            print(warning_message)
+            print("G does not appear to have multiple edges between any " +
+            "pair of nodes. No information should be lost in the switch to " +
+            "DiGraph. Proceeding with flow calculation.\n")
+            H = nx.DiGraph(G)
+        else:
+            H=G
+    else:
+        print("Provided graph is not a MultiDiGraph or DiGraph and other \
+        errors may occur. Abort.\n")
+        return None, None, G
+
+    max_flow_path = nx.maximum_flow(H, s, t)
+    max_flow = nx.maximum_flow_value(H, s, t)
+    return max_flow_path, max_flow, H
+
+
+def find_pathway_flow(G, pathway):
+    '''returns the maximum permissible flow for a given pathway. Any edge
+    without 'capacity' attribute will be given infinite capacity.
+    '''
+    multiedges = has_multiedges(G)
+    if multiedges != True:
+        # graph either has no multiedges, or is not in MultiDiGraph format
+        H = nx.DiGraph(G)
+        edges = set(pairwise(pathway))
+        #H.remove_nodes_from([n for n in G if n not in pathway])
+        H_sg = H.edge_subgraph(edges).copy()
+
+        sources = get_sources(H_sg)
+        sinks = get_sinks(H_sg)
+
+        # check for capacity in edges (if no capacity, NetworkX will raise an
+        # exception) and if every edge has infinite flow, return inf
+        infinite_flow = True
+        for edge in edges:
+            if 'capacity' not in H_sg.get_edge_data(*edge):
+                return math.inf
+            if math.isinf(H_sg.get_edge_data(*edge)['capacity']) is False:
+                infinite_flow = False
+            
+        if infinite_flow is True:
+            return math.inf
+        else:
+            path_flow = nx.maximum_flow_value(H_sg, sources[0], sinks[0])
+            return path_flow
+
+    else:
+        print("Provided pathway has multiple edges. Flow calculation across \
+               multi-edged pathways is currently unsupported. Returning None")
+        return None
+
+
+def find_simple_cycles(G): # pragma: no cover
+    '''finds cycles in a graph and returns them in a list of lists.
+    '''
+    sc = list(nx.simple_cycles(G))
+    return sc
+
+
+def check_if_sublist(path, set_of_steps):
+    '''Checks to see if a pathway contains each element in steps (list)
+    in order. Returns True/False and the position where the steps begin. If
+    False, returns position -1.
+    '''
+    pos = -1
+    sub_list = False
+
+    for i in range(len(path) - len(set_of_steps)+1):
+        if path[i] == set_of_steps[0]:
+            n = 1
+            while (n < len(set_of_steps) and path[i+n] == set_of_steps[n]):
+                n+=1
+            if n == len(set_of_steps):
+                sub_list = True
+                pos = i + len(set_of_steps)
+                break
```

Comment:
Does this also do what you want? (might need to double check the "+1" logic)
```suggestion
num_steps = len(set_of_steps)

start = 0
end = start + num_steps + 1
while (end < len(path)) and (path[start:end] != set_of_steps):
    start += 1
    end = start + num_steps + 1
```

Reply:
I guess with this, you still need to implement an additional check to confirm if the while loop broke because it hit the nd of the path or because it found the full sublist. How about this instead?

```suggestion
        num_steps = len(list_of_steps)
    start = 0
    end = start + num_steps
    while end <= len(path):
        if path[start:end] == tuple(list_of_steps):
            return True, start
        start += 1
        end = start + num_steps
        
    return False, -1
```